src/translate-virtual-address.h

```diff
@@ -88,49 +88,59 @@ static inline bool read_ram_uint64(STATE_ACCESS &a, uint64_t paddr, uint64_t *pv
     return true;
 }
 
-/// \brief Walk the page table and translate a virtual address to the corresponding physical address
-/// \tparam STATE_ACCESS Class of machine state accessor object.
-/// \tparam UPDATE_PTE Whether PTE entries can be modified during the translation.
-/// \param a Machine state accessor object.
-/// \param vaddr Virtual address
-/// \param ppaddr Pointer to physical address.
-/// \param xwr_shift Encodes the access mode by the shift to the XWR triad (PTE_XWR_R_SHIFT,
-///  PTE_XWR_R_SHIFT, or PTE_XWR_R_SHIFT)
-/// \details This function is outlined to minimize host CPU code cache pressure.
-/// \returns True if succeeded, false otherwise.
-template <typename STATE_ACCESS, bool UPDATE_PTE = true>
-static NO_INLINE bool translate_virtual_address(STATE_ACCESS &a, uint64_t *ppaddr, uint64_t vaddr, int xwr_shift) {
-    auto priv = a.read_iflags_PRV();
-    uint64_t mstatus = a.read_mstatus();
-
-    // When MPRV is set, data loads and stores use privilege in MPP
-    // instead of the current privilege level (code access is unaffected)
-    if (xwr_shift != PTE_XWR_X_SHIFT && (mstatus & MSTATUS_MPRV_MASK)) {
-        priv = (mstatus & MSTATUS_MPP_MASK) >> MSTATUS_MPP_SHIFT;
+namespace details {
+template <uint8_t TRANSLATION_MODE, uint8_t ACCESS_TYPE>
+static uint8_t to_mcause_code() {
+    if constexpr (TRANSLATION_MODE == TRANSLATION_G) {
+        if constexpr (ACCESS_TYPE == ACCESS_TYPE_LOAD) return MCAUSE_LOAD_GUEST_PAGE_FAULT;
+        else if constexpr (ACCESS_TYPE == ACCESS_TYPE_STORE) return MCAUSE_STORE_AMO_GUEST_PAGE_FAULT;
+        else return MCAUSE_INSTRUCTION_GUEST_PAGE_FAULT;
+    } else {
+        if constexpr (ACCESS_TYPE == ACCESS_TYPE_LOAD) return MCAUSE_LOAD_PAGE_FAULT;
+        else if constexpr (ACCESS_TYPE == ACCESS_TYPE_STORE) return MCAUSE_STORE_AMO_PAGE_FAULT;
+        else return MCAUSE_FETCH_PAGE_FAULT;
+    }
+}
+
+template <typename STATE_ACCESS, uint8_t TRANSLATION_MODE, uint8_t ACCESS_TYPE, bool UPDATE_PTE>
+static uint8_t do_translate_virtual_address(STATE_ACCESS &a, uint64_t *ppaddr, uint64_t vaddr, int xwr_shift, uint8_t priv) {
+    bool mxr = false;
+    bool sum = false;
+    uint64_t atp = 0;
+    uint8_t widenbits = 0;
+    if constexpr (TRANSLATION_MODE == TRANSLATION_HS) {
+        mxr = a.read_mstatus() & MSTATUS_MXR_MASK;
+        sum = a.read_mstatus() & MSTATUS_SUM_MASK;
+        atp = a.read_satp();
+    } else if constexpr (TRANSLATION_MODE == TRANSLATION_G) {
+        mxr = a.read_mstatus() & MSTATUS_MXR_MASK;
+        atp = a.read_hgatp();
```

Comment:
Well, I think is not necessary after all, because I noticed now that `hgatp` can only be changed when virtual mode is off, and when virtual mode is off the address translation function will not use it. And enabling virtual mode will trigger a TLB shootdown.

But be aware of these details, maybe `write_csr_hgatp` deserves a comment if you understood.